React api translation (#30)
reactjs/pl.reactjs.org: content/docs/reference-react.md

---
id: react-api

> title: API Reacta

Reviewer: Skoro mowa o "Top-Level", to może "Główne API Reacta"?

  - "docs/top-level-api-zh-CN.html"
---

> `React` jest podstawowym elementem biblioteki reaktowej. Jeśli załadujesz go używając tagu `<script>`, główne API `Reacta` będą dostępne globalnie. Jeśli używasz standardu ES6, możesz korzystać ze składni `import React from 'react'`. Jeśli używasz starszego standardu ES5, powinieneś użyć składni `var React = require('react')`.

Reviewer: ```suggestion
Obiekt `React` jest punktem wejściowym do biblioteki React. Jeśli załadujesz ją używając tagu `<script>`, główny interfejs API będzie dostępny w zmiennej globalnej `React`. Jeśli używasz standardu ES6, możesz skorzystać ze składni `import React from 'react'`. Jeśli używasz starszego standardu ES5, użyj składni `var React = require('react')`.
```

> ## Przegląd {#overview}

Reviewer: ```suggestion
## Ogólne informacje {#overview}
```

### Komponenty {#components}

> Komponenty reaktowe pozwalają podzielić interfejs użytkownika na niezależne, enkapsulowane elementy, z których możesz korzystać w wielu miejscach. Komponenty w Reakcie mogą być definiowane dziedziczeniem po klasie `React.Component` lub `React.PureComponent`.

Reviewer: ```suggestion
Komponenty reaktowe pozwalają podzielić interfejs użytkownika na niezależne, enkapsulowane elementy, z których możesz korzystać w wielu miejscach. Komponenty w Reakcie mogą być definiowane poprzez dziedziczenie po klasie `React.Component` lub `React.PureComponent`.
```

 - [`React.Component`](#reactcomponent)
 - [`React.PureComponent`](#reactpurecomponent)

> Jeśli nie korzystasz z klas ES6, możesz skorzystać z modułu `create-react-class`. O korzystaniu z Reacta bez ES6, możesz [przeczytać tutaj](/docs/react-without-es6.html).

Reviewer: ```suggestion
Jeśli nie używasz standardu ES6 do definiowania klas, możesz skorzystać z modułu `create-react-class`. O tym, jak skorzystać z Reacta bez ES6 napisaliśmy w [osobnym rozdziale](/docs/react-without-es6.html) .
```

> Komponenty reaktowe mogą być też definiowane jako funkcje, które mogą być otaczane:

Reviewer: ```suggestion
Komponenty reactowe można również zdefiniować jako funkcje, a następnie otoczyć:
```

- [`React.memo`](#reactmemo)

> ### Tworzenie elementów reaktowych {#creating-react-elements}

Reviewer: ```suggestion
### Tworzenie elementów reactowych {#creating-react-elements}
```

> Polecamy [korzystanie z JSX](/docs/introducing-jsx.html), by opisać wygląd elementów interfejsu użytkownika. Każdy element JSX-owy to wygodniejsza reprezentacja wywołania metody [`React.createElement()`](#createelement). Jeśli korzystasz z JSX, prawdopodobnie nie będziesz musiał korzystać z tej metody.

Reviewer: ```suggestion
Do opisania wyglądu interfejsu użytkownika polecamy [korzystanie ze składni JSX](/docs/introducing-jsx.html). W JSX elementy są tylko wygodniejszymi odpowiednikami wywołania metody [`React.createElement()`](#createelement). Jeśli korzystasz z JSX, prawdopodobnie nigdy nie przyjdzie ci korzystać z tej metody bezpośrednio.
```

- [`createElement()`](#createelement)
- [`createFactory()`](#createfactory)

> Po więcej informacji zobacz [React bez JSX](/docs/react-without-jsx.html).

Reviewer: ```suggestion
Aby dowiedzieć się więcej na ten temat, zajrzyj do rozdziału pt. ["React bez JSX"](/docs/react-without-jsx.html).
```

### Przekształcanie elementów {#transforming-elements}

> `React` udostępnia kilka różnych API do manipulowania elementami:

Reviewer: ```suggestion
`React` udostępnia kilka API do manipulowania elementami:
```

- [`cloneElement()`](#cloneelement)
- [`isValidElement()`](#isvalidelement)
- [`React.Children`](#reactchildren)

### Fragmenty {#fragments}

> `React` pozwala także renderować wiele komponentów, bez komponentu otaczającego przy użyciu:

Reviewer: ```suggestion
`React` pozwala także renderować wiele komponentów bez konieczności używania komponentu opakowującego.
```

- [`React.Fragment`](#reactfragment)

> ### Odniesienia *(ang. Refs)* {#refs}

Reviewer: ```suggestion
### Referencje (ang. *refs*) {#refs}
```

- [`React.createRef`](#reactcreateref)
- [`React.forwardRef`](#reactforwardref)

> ### Suspens *(ang. Suspense)* {#suspense}

Reviewer: ```suggestion
### Suspense {#suspense}
```
Wolałbym tego nie tłumaczyć lub nazwać czymś bardziej neutralnym, np. "wstrzymanie".

> `Suspense` pozwala komponentowi poczekać na konkretne zdarzenie przez wyrenderowaniem się. W tej chwili `Suspense` wspiera jedynie jeden przypadek użycia: [dynamiczne ładowanie komponentów przy użyciu `React.lazy`](/docs/code-splitting.html#reactlazy). W przyszłości będzie wspierał także inne przypadki użycia jak na przykład pobieranie danych zewnętrznych.

Reviewer: ```suggestion
`Suspense` pozwala komponentowi "poczekać" na konkretne zdarzenie przez wyrenderowaniem. Obecnie `Suspense` obsługuje tylko jeden przypadek użycia: [dynamiczne ładowanie komponentów przy użyciu `React.lazy`](/docs/code-splitting.html#reactlazy). W przyszłości będzie wspierał także inne przypadki użycia, jak na przykład pobieranie danych.
```

- [`React.lazy`](#reactlazy)
- [`React.Suspense`](#reactsuspense)

### Hooki {#hooks}

> *Hooki* są nowym dodatkiem do Reacta 16.8. Pozwalają Ci na korzystanie ze stanu oraz innych cech Reacta bez używania klas. Hooki mają [dedykowany rozdział w dokumentacji](/docs/hooks-intro.html) oraz osobne API:

Reviewer: ```suggestion
*Hooki* są nowym dodatkiem do Reacta 16.8. Pozwalają na korzystanie ze stanu oraz innych funkcjonalności Reacta bez konieczności definiowania klas. Hooki mają [dedykowany rozdział w dokumentacji](/docs/hooks-intro.html) oraz osobny interfejs API:
```

* * *

> ## Odwołania {#reference}

Reviewer: ```suggestion
## Dokumentacja {#reference}
```

### `React.Component` {#reactcomponent}

> `React.Component` to podstawowy rodzaj komponentu reaktowego zdefiniowanego przez [klasy ES6](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes):

Reviewer: ```suggestion
`React.Component` to podstawowy rodzaj komponentu reaktowego, definiowany przez użycie [klasy ze standardu ES6](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes):
```

```javascript
class Greeting extends React.Component {
  render() {
    return <h1>Witaj, {this.props.name}</h1>;
  }
}
```

> Zobacz listę metod i atrybutów w odniesieniu do `React.Component` w [API React.Component](/docs/react-component.html).

Reviewer: ```suggestion
Lista metod i właściwości związanych z klasą `React.Component` znajduje się w [dokumentacji API dla React.Component](/docs/react-component.html).
```

* * *

### `React.PureComponent` {#reactpurecomponent}

> `React.PureComponent` jest podobny do [`React.Component`](#reactcomponent). Różnica między nimi jest taka, że [`React.Component`](#reactcomponent) nie implementuje automatycznie [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate), natomiast `React.PureComponent` implementuje go z płytkim porównaniem atrybutów i stanu.

Reviewer: ```suggestion
Klasa `React.PureComponent` jest podobna do [`React.Component`](#reactcomponent). Różnica między nimi jest taka, że [`React.Component`](#reactcomponent) nie posiada implementacji metody [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate), natomiast `React.PureComponent` implementuje ją z użyciem płytkiego porównania atrybutów i stanu.
```

Jeśli metoda `render()` danego komponentu renderuje ten sam rezultat przy tych samych atrybutach i stanie, możesz przekształcić go na `React.PureComponent`, by poprawić wydajność.

> Uwaga
>

> > Metoda `shouldComponentUpdate()` klasy `React.PureComponent` wykonuje jedynie płytkie porównanie obiektów. Jeśli zawierają one skomplikowaną strukturę danych, może to wprowadzić nieprawidłowości podczas zmian na głębszych poziomach struktury. Korzystaj z `PureComponent` tylko, jeśli spodziewasz się prostej struktury atrybutów i stanu, lub użyj [`forceUpdate()`](/docs/react-component.html#forceupdate), jeśli wiesz, że dane umieszczone w strukturze głęboko uległy zmianie. Możesz także zastanowić się nad skorzystaniem z [niezmiennych obiektów (ang. immutable objects)](https://facebook.github.io/immutable-js/), by ułatwić szybkie porównanie zagnieżdżonych danych.

Reviewer: ```suggestion
> Metoda `shouldComponentUpdate()` klasy `React.PureComponent` wykonuje jedynie płytkie porównanie obiektów. Jeśli zawierają one skomplikowaną strukturę danych, może to wprowadzić nieprawidłowości podczas zmian na głębszych poziomach struktury. Korzystaj z `PureComponent` tylko, jeśli spodziewasz się prostej struktury atrybutów i stanu, lub użyj [`forceUpdate()`](/docs/react-component.html#forceupdate), jeśli wiesz, że dane umieszczone w strukturze głęboko uległy zmianie. Możesz także zastanowić się nad skorzystaniem z [niezmiennych obiektów (ang. *immutable objects*)](https://facebook.github.io/immutable-js/), by ułatwić szybkie porównanie zagnieżdżonych danych.
```

>

> > Dodatkowo, `shouldComponentUpdate()` klasy `React.PureComponent` pomija aktualizację atrybutów dla całego poddrzewa komponentu. Upewnij się, że wszystkie dzieci również korzystają z tego rozwiązania.

Reviewer: ```suggestion
> Dodatkowo, `shouldComponentUpdate()` klasy `React.PureComponent` pomija aktualizację atrybutów dla całego poddrzewa komponentu. Upewnij się, że wszystkie komponenty potomne również korzystają z tego rozwiązania.
```

* * *

### `React.memo` {#reactmemo}

```javascript
const MyComponent = React.memo(function MyComponent(props) {
  /* renderuj korzystając z atrybutów */
});
```

> `React.memo` jest [komponentem wyższego rzędu](/docs/higher-order-components.html). Jego działanie jest podobne do [`React.PureComponent`](reactpurecomponent), jednak stosowany jest tylko do komponentów funkcyjnych.

Reviewer: ```suggestion
`React.memo` jest [komponentem wyższego rzędu (ang. *higher-order component*](/docs/higher-order-components.html). Jego działanie jest podobne do [`React.PureComponent`](#reactpurecomponent), jednak stosowany jest tylko do komponentów funkcyjnych.
```

> Jeśli twój komponent funkcyjny renderuje ten sam rezultat, dostając te same atrybuty, możesz opakować go w `React.memo` by poprawić wydajność. Oznacza to, że React pominie renderowanie tego komponentu i użyje jego ostatnio zrenderowaną wersję.

Reviewer: ```suggestion
Jeśli twój komponent funkcyjny przy takich samych atrybutach zawsze renderuje tę samą strukturę, możesz opakować go w `React.memo` w celu poprawy wydajności. Oznacza to, że React pominie renderowanie tego komponentu i użyje jego ostatnio wyrenderowanej wersji.
```

> Domyślnie, komponent wykona jedynie płytkie porównanie obiektów w atrybutach. Jeśli chcesz kontrolować renderowanie komponentu, możesz dostarczyć własną funkcję do porównywania jako drugi argument.

Reviewer: ```suggestion
Domyślnie, komponent wykona jedynie płytkie porównanie obiektów przekazanych w atrybutach. Jeśli chcesz zastosować własny mechanizm porównujący, możesz przekazać odpowiednią funkcję jako drugi argument.
```

```javascript
function MyComponent(props) {
  /* renderuj korzystając z atrybutów */
}
function areEqual(prevProps, nextProps) {
  /*

>   zwróć true jeśli przekazując nextProps do rendera zwróciłoby

Reviewer: ```suggestion
  zwróć true, jeśli przekazując nextProps, komponent zwróciłby 
```

>   taki sam rezultat jak podczas przekazania prevProps,

Reviewer: ```suggestion
  taki sam rezultat, jak po przekazaniu prevProps,
```

  w innym przypadku zwróć false
  */
}
export default React.memo(MyComponent, areEqual);
```

> Ta metoda wykorzystywana jest jedynie do **[optymalizacji wydajnośći](/docs/optimizing-performance.html).** Nie powinieneś polegać na niej do "zapobiegania" renderowaniu. Może to jedynie doprowadzić do większej ilości błędów.

Reviewer: ```suggestion
Ta metoda wykorzystywana jest jedynie do **[optymalizacji wydajności](/docs/optimizing-performance.html).** Nie powinno się używać jej do "zapobiegania" renderowaniu, ponieważ może to doprowadzić do większej ilości błędów.
```

> Uwaga
>

> > W odróżnieniu od metody [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate) w komponencie klasowym, funkcja `areEqual` zwraca `true` jeśli atrybuty są jednakowe lub `false` jeśli nie są. Jest to odwrócona logika metody `shouldComponentUpdate`.

Reviewer: ```suggestion
> W odróżnieniu od metody [`shouldComponentUpdate()`](/docs/react-component.html#shouldcomponentupdate) w komponencie klasowym, funkcja `areEqual` zwraca `true`, jeśli atrybuty są jednakowe, lub `false`, jeśli nie są. Jest to odwrócona logika metody `shouldComponentUpdate`.
```

)
```

> Tworzy i zwraca nowy [element reaktowy](/docs/rendering-elements.html) danego typu. Argument `type` może być zarówno nazwą tagu HTML-owego (np. `'div'` lub `'span'`), typem [komponentu reaktowego](/docs/components-and-props.html) (klasa lub funkcja) jak i typem [fragmentu reaktowego](#reactfragment).

Reviewer: ```suggestion
Tworzy i zwraca nowy [element reactowy](/docs/rendering-elements.html) danego typu. Argument `type` może być zarówno nazwą znacznika HTML (np. `'div'` lub `'span'`), [komponentem reaktowym](/docs/components-and-props.html) (klasa lub funkcja), jak i [fragmentem reaktowym](#reactfragment).
```

> Kod pisany w [JSX](/docs/introducing-jsx.html) jest konwertowany do `React.createElement()`. Jeśli korzystasz z JSX, w większości przypadków nie będziesz bezpośrednio wywoływał `React.createElement()`. Zobacz [React bez JSX](/docs/react-without-jsx.html), by dowiedzieć się więcej.

Reviewer: ```suggestion
Kod pisany w [JSX](/docs/introducing-jsx.html) jest konwertowany do wywołań funkcji `React.createElement()`. Jeśli korzystasz z JSX, w większości przypadków nie będziesz bezpośrednio wywoływać `React.createElement()`. Przeczytaj rozdział pt. ["React bez JSX"](/docs/react-without-jsx.html), aby dowiedzieć się więcej.
```

)
```

> Zacznij od sklonowania i zwrócenia elementu reaktowego, przy pomocy `element`. Nowo utworzony element będzie posiadał początkowe atrybuty połączone płytkim porównaniem z nowymi atrybutami. Nowe dzieci zastąpią obecne. `key` oraz `ref` z oryginalnego elementu zostaną zachowane.

Reviewer: ```suggestion
Zacznij od sklonowania i zwrócenia elementu reactowego przy pomocy funkcji `element`. Nowo utworzony element będzie posiadał pierwotne atrybuty połączone płytko z nowymi. Nowe elementy potomne zastąpią obecne. `key` oraz `ref` z pierwotnego elementu zostaną zachowane.
```

> `React.cloneElement()` jest niemal identyczny z:

Reviewer: ```suggestion
Wywołanie `React.cloneElement()` jest niemal równoznaczne z napisaniem:
```

```js
<element.type {...element.props} {...props}>{children}</element.type>
```

> Jednakże zachowuje on także wszystkie `ref`. Oznacza to, że jeśli jedno z dzieci posiada `ref`, nie zostanie ono przypadkowo zabrane z jednego z przodków. Nowy element będzie posiadał ten sam `ref`.

Reviewer: ```suggestion
Jednakże zachowuje on także wszystkie `ref`. Oznacza to, że jeśli jeden z komponentów potomnych posiada `ref`, nie zostanie on przypadkowo zabrany z jednego z przodków. Nowy element będzie posiadał ten sam `ref` co przed klonowaniem.
```

> API to zostało przygotowane w zastępstwie za przestarzałe `React.addons.cloneWithProps()`.

Reviewer: Według mnie trochę zbyt dosłownie przetłumaczone. Może:
```suggestion
Powyższy interfejs API zastąpił przestarzałą funkcję `React.addons.cloneWithProps()`.
```

React.createFactory(type)
```

> Zwraca funkcję, która tworzy reaktowy element danego typu. Tak jak [`React.createElement()`](#createElement), argument `type`, może być zarówno nazwą tagu HTML-owego (np. `'div'` lub `'span'`), typem [komponentu reaktowego](/docs/components-and-props.html) (klasa lub funkcja) lub typem [fragmentu reaktowego](#reactfragment).

Reviewer: ```suggestion
Zwraca funkcję, która tworzy reaktowy element danego typu. Tak jak [`React.createElement()`](#createElement), argument `type`, może być zarówno nazwą znacznika HTML (np. `'div'` lub `'span'`), [komponentem reactowym](/docs/components-and-props.html) (klasa lub funkcja) lub [fragmentem reactowym](#reactfragment).
```

Ta funkcja pomocnicza jest uznawana za przestarzałą i radzimy korzystać z JSX lub bezpośrednio z `React.createElement()`.

> Jeśli korzystasz z JSX, nie ma potrzeby korzystania z `React.createFactory()`. Zobacz [React bez JSX](/docs/react-without-jsx.html), by dowiedzieć się więcej.

Reviewer: ```suggestion
Jeśli korzystasz z JSX, nie ma potrzeby korzystania z `React.createFactory()`. Przeczytaj rozdział pt. ["React bez JSX"](/docs/react-without-jsx.html), aby dowiedzieć się więcej.
```

React.isValidElement(object)
```

> Weryfikuje czy obiekt jest elementem reaktowym. Zwraca `true` lub `false`.

Reviewer: ```suggestion
Weryfikuje, czy obiekt jest elementem reactowym. Zwraca `true` lub `false`.
```

* * *

### `React.Children` {#reactchildren}

> `React.Children` udostępnia narzędzia do obsługi struktury danych z `this.props.children`.

Reviewer: ```suggestion
`React.Children` udostępnia narzędzia do obsługi struktury danych przekazanej w `this.props.children`.
```

#### `React.Children.map` {#reactchildrenmap}

```javascript
React.Children.map(children, function[(thisArg)])
```

> Wywołuje funkcję na każdym bezpośrednim dziecku, zawartym w `children` z `this` ustawionym na `thisArg`. Jeśli `children` jest tablicą, funkcja zostanie wywołana na każdym dziecku zawartym w tej tablicy.

Reviewer: ```suggestion
Wywołuje funkcję na każdym bezpośrednim komponencie potomnych zawartym w `children`, z `this` ustawionym na `thisArg`. Jeśli `children` jest tablicą, funkcja zostanie wywołana na każdym elemencie tej tablicy.
```

> Uwaga
>

> > Jeśli `children` jest `Fragmentem`, zostanie potraktowana jako pojedyncze dziecko, nie jak tablica.

Reviewer: ```suggestion
> Jeśli `children` jest typu `Fragment`, zostanie potraktowany jako pojedynczy potomek, nie jak tablica.
```

#### `React.Children.forEach` {#reactchildrenforeach}

```javascript
React.Children.forEach(children, function[(thisArg)])
```

> Działa tak samo jak [`React.Children.map()`](#reactchildrenmap), ale nie zwraca tablicy.

Reviewer: ```suggestion
Działa tak samo, jak [`React.Children.map()`](#reactchildrenmap), ale nie zwraca tablicy.
```

#### `React.Children.count` {#reactchildrencount}

```javascript
React.Children.count(children)
```

> Zwraca liczbę komponentów w `children` równą ilości razy, jakie byłby wywołany callback przekazany do `map` lub `forEach`.

Reviewer: ```suggestion
Zwraca liczbę komponentów w `children`, równą liczbie potencjalnych wywołań funkcji zwrotnej (ang. *callback*) przekazanej do `map` lub `forEach`.
```

#### `React.Children.only` {#reactchildrenonly}

```javascript
React.Children.only(children)
```

> Weryfikuje czy `children` ma tylko jedno dziecko (element reaktowy) i zwraca je. W innym przypadku metoda zwróci błąd.

Reviewer: ```suggestion
Weryfikuje, czy `children` ma tylko jednego potomka (element reactowy), i zwraca go. W innym przypadku metoda rzuci wyjątkiem.
```

> Uwaga
>

> >`React.Children.only()` nie akceptuje wartości zwracanej przez [`React.Children.map()`](#reactchildrenmap), ponieważ jest to tablica, a nie element reaktowy

Reviewer: ```suggestion
>`React.Children.only()` nie akceptuje wartości zwracanej przez [`React.Children.map()`](#reactchildrenmap), ponieważ jest to tablica, a nie element reactowy.
```

#### `React.Children.toArray` {#reactchildrentoarray}

```javascript
React.Children.toArray(children)
```

> Zwraca strukturę danych z `children` jako płaską tablicę z kluczem przypisanym do każdego z dzieci. Jest to przydatne, jeśli chcemy manipulować grupą dzieci w metodzie renderowania. Szczególnie jeśli chcemy zmienić kolejność lub podzielić `this.props.children`, zanim przekażemy je dalej.

Reviewer: ```suggestion
Zwraca strukturę z `children` jako płaską tablicę z kluczem przypisanym do każdego z potomków. Przydatne, jeśli chcemy manipulować grupą potomków w metodzie renderującej, zwłaszcza jeśli chcemy zmienić ich kolejność lub podzielić przed przekazaniem dalej.
```

> Uwaga
>

> > `React.Children.toArray()` zmienia klucze, by zachować semantyczną poprawność zagłębionych tablic podczas przygotowywania płaskiej struktury. Oznacza to, że `toArray` rozpocznie nazwę każdego klucza, tak by każdy element należał do odpowiedniego zakresu.

Reviewer: ```suggestion
> `React.Children.toArray()` zmienia klucze, by zachować semantyczną poprawność zagłębionych tablic podczas przygotowywania płaskiej struktury. Oznacza to, że `toArray` doda prefiks do nazwy każdego klucza, tak by każdy element należał do odpowiedniego zakresu.
```

* * *

### `React.Fragment` {#reactfragment}

Komponent `React.Fragment` pozwala zwrócić wiele elementów w metodzie `render()` bez opakowywania ich w dodatkowy element DOM:

```javascript
render() {
  return (
    <React.Fragment>
      Jakiś tekst.
      <h2>Nagłowek</h2>
    </React.Fragment>
  );
}
```

Możesz także skorzystać ze skróconego zapisu `<></>`. Po więcej informacji zajrzyj do wpisu na blogu - [React v16.2.0: Improved Support for Fragments](/blog/2017/11/28/react-v16.2.0-fragment-support.html).

### `React.createRef` {#reactcreateref}

> `React.createRef` tworzy [referencję](/docs/refs-and-the-dom.html), którą możesz przypiąć do dowolnego elementu reaktowego poprzez atrybut ref.

Reviewer: ```suggestion
`React.createRef` tworzy [referencję](/docs/refs-and-the-dom.html), którą możesz przypiąć do dowolnego elementu reactowego poprzez atrybut `ref`.
```

`embed:16-3-release-blog-post/create-ref-example.js`

### `React.forwardRef` {#reactforwardref}

> `React.forwardRef` tworzy komponent reaktowy, który przekazuje atrybut [ref](/docs/refs-and-the-dom.html) do kolejnego komponentu w dół drzewa. Ta technika nie jest zbyt popularna, ale przydatna w dwóch konkretnych przypadkach:

Reviewer: ```suggestion
`React.forwardRef` tworzy komponent reactowy, który przekazuje atrybut [`ref`](/docs/refs-and-the-dom.html) do kolejnego komponentu w dół drzewa. Ta technika nie jest zbyt popularna, ale przydatna w dwóch konkretnych przypadkach:
```

* [Przekazywanie referencji komponentom DOM](/docs/forwarding-refs.html#forwarding-refs-to-dom-components)
* [Przekazywanie referencji komponentom wyższego rzędu](/docs/forwarding-refs.html#forwarding-refs-in-higher-order-components)

> `React.forwardRef` przyjmuje funkcję renderującą jako argument. React wywoła tę funkcję biorąc `props` oraz `ref` jako dwa argumenty. Funkcja ta powinna zwrócić węzeł reaktowy.

Reviewer: ```suggestion
`React.forwardRef` przyjmuje funkcję renderującą jako argument. React wywoła tę funkcję, przekazując `props` oraz `ref` jako argumenty. Funkcja ta powinna zwrócić węzeł reactowy.
```

`embed:reference-react-forward-ref.js`

> W powyższym przykładzie, React przekazuje `ref` będąc dla `<FancyButton ref={ref}>` drugim argumentem do funkcji renderującej wewnątrz `React.forwardRef`. Funkcja renderująca przekaże `ref` do elementu `<button ref={ref}>`.

Reviewer: ```suggestion
W powyższym przykładzie React przekazuje `ref` poprzez `<FancyButton ref={ref}>` jako drugi argument funkcji renderującej wewnątrz `React.forwardRef`. Funkcja renderująca przekaże `ref` do elementu `<button ref={ref}>`.
```

> W rezultacie po tym, jak referencja zostanie przypięta przez Reacta, `ref.current` będzie wskazywał bezpośrednio na element `<button>`.

Reviewer: ```suggestion
W rezultacie, po tym, jak referencja zostanie przypięta przez Reacta, `ref.current` będzie wskazywał bezpośrednio na element `<button>`.
```

> Po więcej informacji zobacz [przekazywanie referencji](/docs/forwarding-refs.html).

Reviewer: ```suggestion
Aby dowiedzieć się więcej, przeczytaj rozdział poświęcony [przekazywaniu referencji](/docs/forwarding-refs.html).
```

### `React.lazy` {#reactlazy}

> `React.lazy()` pozwala Ci zdefiniować komponenty, które są ładowane dynamicznie. Pozwala to zredukować rozmiar paczki, poprzez wyodrębnienie komponentów, które nie są używane podczas inicjalizacji renderowania.

Reviewer: ```suggestion
`React.lazy()` pozwala definiować komponenty, które są ładowane dynamicznie. Zmniejsza się w ten sposób rozmiar paczki, ponieważ wyodrębnione zostają komponenty, które nie są używane podczas wstępnego renderowania.
```

> W naszej [dokumentacji poświęconej rozdzielaniu kodu](/docs/code-splitting.html#reactlazy) zamieściliśmy więcej informacji na ten temat. [Ten artykuł](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d) również może okazać się pomocny w wyjaśnieniu szczegółów działania tej metody.

Reviewer: ```suggestion
W naszej [dokumentacji poświęconej dzieleniu kodu](/docs/code-splitting.html#reactlazy) zamieściliśmy więcej informacji na ten temat. [Ten artykuł](https://medium.com/@pomber/lazy-loading-and-preloading-components-in-react-16-6-804de091c82d) również może okazać się pomocny w wyjaśnieniu zasad działania tej metody.
```

```js
// Ten komponent jest ładowany dynamicznie
const SomeComponent = React.lazy(() => import('./SomeComponent'));
```

> Pamiętaj, że renderowanie leniwych komponentów (ang. *lazy*) wymaga użycia komponentu `<React.Suspense>` wyżej w drzewie renderowania. W ten sposób definiujesz wskaźnik ładowania.

Reviewer: ```suggestion
Pamiętaj, że renderowanie "leniwych komponentów" (ang. *lazy components*) wymaga użycia komponentu `<React.Suspense>` na wyższym poziomie drzewa. W ten sposób definiuje się wskaźnik ładowania.
```
Nie do końca pasuje mi tu ostatnie zdanie, ale w angielskiej wersji też jest to dość niejasno napisane...

> **Uwaga**
>

> > Używanie `React.lazy` z dynamicznymi importami wymaga, by w środowisku JS dostępne były Promisy. Będzie to wymagało użycia polyfilli dla przeglądarek IE11 i starszych.

Reviewer: ```suggestion
> Używanie `React.lazy` z dynamicznymi importami wymaga, by w środowisku JS dostępna była klasa `Promise`. Będzie to wymagało użycia tzw. "łatek" (ang. *polyfill*) dla przeglądarki IE11 i starszych.
```

### `React.Suspense` {#reactsuspense}

> `React.Suspense` pozwala zdefiniować wskaźnik ładowania w razie, gdy któryś z komponentów poniżej nie jest jeszcze gotowy do renderowania. W tej chwili jedynym przypadkiem użycia `<React.Suspense>` jest dynamiczne ładowanie komponentów.

Reviewer: ```suggestion
`React.Suspense` pozwala zdefiniować wskaźnik ładowania, w razie gdyby któryś z komponentów poniżej nie b ył jeszcze gotowy do wyrenderowania. Obecnie jedynym przypadkiem użycia `<React.Suspense>` jest dynamiczne ładowanie komponentów.
```

}
```

> W naszej [dokumentacji poświęconej rozdzielaniu kodu](/docs/code-splitting.html#reactlazy) zamieściliśmy więcej informacji na ten temat. Zwróć uwagę na to, że komponenty `lazy` mogą być zawarte głęboko poniżej `Suspense`. Nie musisz otaczać każdego elementu w `<Suspense>`, jednak musisz użyć `lazy()` dla każdego komponentu, którego chcesz użyć.

Reviewer: ```suggestion
W naszej [dokumentacji poświęconej rozdzielaniu kodu](/docs/code-splitting.html#reactlazy) zamieściliśmy więcej informacji na ten temat. Zwróć uwagę na to, że komponenty `lazy` mogą być zawarte w drzewie wiele poziomów poniżej `Suspense`. Dobrą praktyką jest umieszczanie `<Suspense>` w miejscu, w którym powinien pojawić się wskaźnik ładowania, natomiast `lazy()` w miejscu, w którym chcesz rozdzielić kod.
```

> Mimo, że nie jest to obecnie wspierane, w planach jest wykorzystanie `Suspense` przy pobieraniu danych. Więcej możesz się dowiedzieć z [naszej roadmapy](/blog/2018/11/27/react-16-roadmap.html).

Reviewer: ```suggestion
Mimo że nie jest to obecnie wspierane, w planach jest wykorzystanie `Suspense` przy pobieraniu danych. Więcej możesz dowiedzieć się z [naszego planu działania](/blog/2018/11/27/react-16-roadmap.html).
```
